feat: add tests for filter methods

keroro520 opened the pull request:
I add some test cases for filter methods.
All of these cases pass on Hardhat network, but all fail on Godwoken network 😂

keroro520 committed:
feat: add tests for logs

RetricSu commented on contracts/test/filterMethods.js:
```diff
@@ -0,0 +1,167 @@
+
+        const producer = 3;
+        const repeats = 3;
+        (await contract.produce(producer, repeats)).wait();
```
```suggestion
        await (await contract.produce(producer, repeats)).wait(2);
```

need to wait for 2 blocks? otherwise, I don't think the database has been updated

ShookLyngs replied:
I thought `wait(2)` was no longer needed?

keroro520 replied:
Oh yes, I need to wait for 2 blocks. `wait(2)` is good for me. Thanks!

RetricSu replied:
> I thought `wait(2)` was no longer needed?

we can get rid of this after this pr https://github.com/nervosnetwork/godwoken-web3/pull/334

most of the time wait(1)(not on chain yet) is enough, but logs require database updates, so need to wait(2)(on chain)

RetricSu commented on contracts/test/filterMethods.js:
```diff
@@ -0,0 +1,167 @@
+        const address = contract.address;
+        const filterId = await ethers.provider.send("eth_newFilter", [{address, topics}]);
+        const tx = await contract.produce(producer, repeats);
+        tx.wait();
```
```suggestion
        await tx.wait(2);
```

same here

keroro520 committed:
feat: enable Hardhat interval mining mode

- This change enables interval-mining mode on Hardhat network.
Without this, tx.wait(2) doesn't work(block waiting forever), as Hardhat won't mine new blocks when no new transactions sending.

  Ref: https://hardhat.org/hardhat-network/docs/explanation/mining-modes

RetricSu commented on contracts/hardhat.config.js:
```diff
@@ -60,7 +60,11 @@ module.exports = {
     hardhat: {
-      hardfork: "berlin"
+      hardfork: "berlin",
+      mining: {
```
why do we need this?

keroro520 replied:
This change enables interval-mining mode on Hardhat network.
Without this, `tx.wait(2)` doesn't work(block waiting forever), as Hardhat won't mine new blocks when no new transactions sending.

keroro520 replied:
Ref: https://hardhat.org/hardhat-network/docs/explanation/mining-modes

RetricSu replied:
I see, we should also remove this after instant-finality is removed.

RetricSu commented on contracts/test/logs.js:
```diff
@@ -0,0 +1,182 @@
+        const numberOfLogs = 2;
+        const topics = [
+            null,
+            "0x" + ("0".repeat(64) + nameOfLogs.toString(16)).slice(-64),
```
manual calculating might not be a good idea, consider using standard methods like https://docs.ethers.io/v5/single-page/#/v5/api/contract/contract/-%23-Contract--filters

keroro520 replied:
So good! Thanks, will update later.

keroro520 replied:
Updated

RetricSu commented on contracts/test/logs.js:
```diff
@@ -0,0 +1,182 @@
+            expect(logs.map(log => Number(log.logIndex)), `filter: ${JSON.stringify(filter)}`).deep.eq(expected);
+        }
+    })
+})
```
```suggestion
})

```

ShookLyngs commented:
New features around `mainnet regression testing` has been merged to the develop branch:
- https://github.com/nervosnetwork/godwoken-tests/pull/153

Since tests on mainnet shouldn't waste too much capacity, right now most of test cases are disabled when testing on mainnet. Based on that, we should consider wether the new test case can be used for mainnet regression testing, and write conditional code for it.

For example, if we think that the new test case shouldn't be used for tests on mainnet right now:
```javascript
const { isGwMainnetV1 } = require('../utils/network');

describe('...', () => {
  if (isGwMainnetV1()) {
    return;
  }

  ...
});
```

keroro520 commented:
> 

Hi @ShookLyngs, thanks for your suggestion.
I will mark this PR draft first, and then make some optimization(these cases cost too much time) and apply your suggestion later.

Flouse commented:
- replaced by https://github.com/godwokenrises/godwoken-tests/pull/238
